### ISSUE TITLE
Writing Custom Adversarial Loss Function

### ISSUE THREAD

**Issue by kevinwu23** [fchollet/keras#6019]
Is it possible to incorporate the output from a discriminator function in Keras?

For example is it at all possible to create something to this effect?

`def adversarial_loss(y_true, y_pred):
    return K.log(1-discriminator.predict(y_pred))
`

where 'discriminator' refers to a model that takes in y_pred and returns a 0 to 1 value.

It appears there have been Tensorflow implementations which combine contextual and perceptual loss:
https://bamos.github.io/2016/08/09/deep-completion/

Any help is greatly appreciated.

**Comment by thefiddler** [comment]
Yes this is definitely possible:

```python
def create_adv_loss(discriminator):
    def loss(y_true, y_pred):
        return K.log(1.0 - discriminator.predict(y_pred))
    return loss

discriminator = Model(...)
model = Model(...)
model.compile(..., loss=create_adversarial_loss(discriminator))
```

**Comment by kevinwu23** [comment]
Thanks for the input.

Have you been able to get this kind of implementation working? I run into issues as 'y_pred' is treated like a Tensor Variable, not an array suitable for passing as a model input.

**Comment by r-kellerm** [comment]
I'm having the same issue - @kevinwu23 did you manage to solve this?

**Comment by kevinwu23** [comment]
Unfortunately using the predict function in a loss function does not work. You will have to append the discriminator to the generator and freeze the discriminator.

**Comment by kevinwu23** [comment]
Yes, if you search GitHub for Keras implementations of GANs, there are a lot of examples. Essentially a frozen discriminator is the same as a loss function.

**Comment by r-kellerm** [comment]
My use case doesn't include a generator to append to the generator. I'm trying to do something similar to the approach described in this paper: [A3T: Adversarially Augmented Adversarial Training](https://arxiv.org/abs/1801.04055) -- connecting two discriminator when one of them tries to fool the other. Is it possible to do that in Keras? 
Thanks!

**Comment by kevinwu23** [comment]
Of course. Just initiate the two discriminators, D1 and D2. 
Then, create M1 = D1 -> D2 and M2 = D2 -> D1 and train M1 and M2 one after the other.

**Comment by r-kellerm** [comment]
Pardon me asking something basic,
What is the meaning of "M1 = D1 -> D2" ? What is the '->' connection? (I guess M1, M2 are Models).
Thanks!

**Comment by kevinwu23** [comment]
It just means you would append the last layer of D1 to the first layer of D2 (And yes, M1 and M2 are models)

**Comment by pranitapradhan91** [comment]
Is there any working code for using predict function in custom loss? Any help is appreciated

**Comment by wanghaichao2195** [comment]
> a frozen discriminator is the same as a loss function.
May I ask why these 2 are the same? I just don't get it. The D and G ought to have differenet loss functions.

**Comment by wanghaichao2195** [comment]
@kevinwu23

**Comment by kevinwu23** [comment]
The output of a discriminator function with BSE loss can be used as your adversarial loss. Keras just doesn't allow for this explicit formulation.
